docs/source/tutorial/introduction.rst

```diff
@@ -0,0 +1,438 @@
+
+You can see here that the indices for the nodes in ``graph`` are missing ``2``.
+Also, after a removal the index of the removed node or edge will be reused on
+subsequent additions. For example, building off the previous example if you ran
+
+.. jupyter-execute::
+
+    graph.add_node("New Node")
+
+this new node is assigned index 2 again.
+
+.. _data_payload:
+
+What to use for node and edge data payload
+==========================================
+
+In the above examples for the most part we use integers, strings, and ``None``
+for the data payload of nodes and edges in graphs (mostly for simplicity).
+However, retworkx allows the use of any Python object as the data payload for
+nodes and edges. This flexibility to use any python object is very powerful as
+it allows you to create graphs that contain other graphs, graphs that contain
+files, graphs with functions, etc. This means you only need to keep a reference
+to the integer index returned by retworkx for the objects you use as a data
+payloads to find those objects in the graph. For example, one approach you can
+take is to store the index as an attribute on the object you add to the graph:
+
+.. jupyter-execute::
+
+    class GraphNode:
+
+        def __init__(self, value):
+            self.value = value
+            self.index = None
+
+    graph = rx.PyGraph()
+    index = graph.add_node(GraphNode("A"))
+    graph[index].index = index
+
+Additionally, at any time you can find the index mapping to the data payload
+and build a mapping or update a reference to it. For example, building on the
+above example you can update the index references all at once after creation:
+
+.. jupyter-execute::
+
+    class GraphNode:
+        def __init__(self, value):
+            self.index = None
+            self.value = value
+
+        def __str__(self):
+            return f"GraphNode: {self.value} @ index: {self.index}"
+
+    class GraphEdge:
+        def __init__(self, value):
+            self.index = None
+            self.value = value
+
+        def __str__(self):
+            return f"EdgeNode: {self.value} @ index: {self.index}"
+
+    graph = rx.PyGraph()
+    graph.add_nodes_from([GraphNode(i) for i in range(5)])
+    graph.add_edges_from([(i, i + 1, GraphEdge(i)) for i in range(4)])
+    # Populate index attribute in GraphNode objects
+    for index in graph.node_indices():
+        graph[index].index = index
+    # Populate index attribute in GraphEdge objects
+    for index, data in graph.edge_index_map().items():
+        data[2].index = index
+    print("Nodes:")
+    for node in graph.nodes():
+        print(node)
+    print("Edges:")
+    for edge in graph.edges():
+        print(edge)
+
+Accessing edges and neighbors
+=============================
+
+You can access edges from a node using the :meth:`~rx.PyGraph.incident_edges`
+method:
+
+.. jupyter-execute::
+
+    print(G.incident_edges(2))
+
+which will return the edge indices of the edges incident a given node. You
```

Comment:
"incident to node 2"

Reply:
Fixed in: https://github.com/Qiskit/retworkx/pull/537/commits/f7e4421111f967d25a0b2303bb34e5e110fdd8cc